### PR TITLE
fix(torch): preserve symbolic compiled shapes

### LINKED ISSUE #145: `make_fx` does not produce a batch-invariant trace for `ewald_reciprocal_space` / `pme_reciprocal_space`
Tracing either reciprocal-space op with `make_fx(..., tracing_mode="symbolic")` at two different batch sizes should give byte-identical .code.
For `ewald_reciprocal_space` this was the case in nvalchemiops 0.4.0 but is no longer the case in nvalchemiops 0.4.1 (regression). For `pme_reciprocal_space` it is not the case in either version, 0.4.0 or 0.4.1.

# Minimum reproducible example:

```python
import difflib

import torch
from torch.fx.experimental.proxy_tensor import make_fx

import nvalchemiops

from nvalchemiops.torch.interactions.electrostatics import (
    generate_k_vectors_ewald_summation,
    generate_k_vectors_pme,
    ewald_reciprocal_space, pme_reciprocal_space
)

print(f"nvalchemiops {nvalchemiops.__version__}, torch {torch.__version__}")
assert torch.cuda.is_available(), "needs a CUDA device (the Warp kernels are GPU-only)"
device = torch.device("cuda")

N_PER_FRAME = 6
DTYPE = torch.float64


def _report(label, code_4, code_5):
    print(f"\n{label}: graphs identical: {code_4 == code_5}")
    if code_4 != code_5:
        diff = [
            line
            for line in difflib.unified_diff(
                code_4.splitlines(), code_5.splitlines(), lineterm="", n=0
            )
            if line[:1] in "-+" and line[:3] not in ("---", "+++")
        ]
        for line in diff[:8]:
            print("   ", line.strip())
        print(f"    ... {len(diff)} differing lines in total")


def make_frames(n_frames):
    """Shared random positions/charges/cell/batch_idx for `n_frames` cubic cells."""
    n = n_frames * N_PER_FRAME
    g = torch.Generator().manual_seed(0)
    pos = (torch.rand(n, 3, generator=g, dtype=DTYPE) * 8.0).to(device)
    q = torch.randn(n, generator=g, dtype=DTYPE)
    q = (q - q.mean()).to(device)
    cell = (torch.eye(3, dtype=DTYPE, device=device) * 8.0).expand(n_frames, 3, 3)
    cell = cell.contiguous()
    batch_idx = torch.arange(n_frames, device=device, dtype=torch.int32)
    batch_idx = batch_idx.repeat_interleave(N_PER_FRAME)
    alpha = torch.full((n_frames,), 0.4, dtype=DTYPE, device=device)
    return pos, q, cell, batch_idx, alpha


def trace_ewald(n_frames):
    pos, q, cell, batch_idx, alpha = make_frames(n_frames)
    k_cutoff = torch.full((n_frames,), 4.0, dtype=DTYPE, device=device)
    k_vectors = generate_k_vectors_ewald_summation(cell=cell, k_cutoff=k_cutoff)

    def f(positions, charges, cells, batch, alphas, kvecs):
        return ewald_reciprocal_space(
            positions=positions,
            charges=charges,
            cell=cells,
            k_vectors=kvecs,
            alpha=alphas,
            batch_idx=batch,
        )

    return make_fx(f, tracing_mode="symbolic", _allow_non_fake_inputs=True)(
        pos, q, cell, batch_idx, alpha, k_vectors
    ).code


MESH_DIMS = (8, 8, 8)

def trace_pme(n_frames):
    pos, q, cell, batch_idx, alpha = make_frames(n_frames)
    k_vectors, k_squared = generate_k_vectors_pme(cell=cell, mesh_dimensions=MESH_DIMS)

    def f(positions, charges, cells, batch, alphas, kvecs, ksq):
        return pme_reciprocal_space(
            positions=positions,
            charges=charges,
            cell=cells,
            alpha=alphas,
            mesh_dimensions=MESH_DIMS,
            batch_idx=batch,
            k_vectors=kvecs,
            k_squared=ksq,
        )

    return make_fx(f, tracing_mode="symbolic", _allow_non_fake_inputs=True)(
        pos, q, cell, batch_idx, alpha, k_vectors, k_squared
    ).code


_report("ewald_reciprocal_space", trace_ewald(4), trace_ewald(5))
_report("pme_reciprocal_space", trace_pme(4), trace_pme(5))
```

# Relevant log outputs:

* nvalchemiops 0.4.0:

```
ewald_reciprocal_space: graphs identical: True

pme_reciprocal_space: graphs identical: False
    -    batch_spline_spread = torch.ops.nvalchemiops.batch_spline_spread.default(positions_1, charges_1, batch_1, transpose, 4, [8, 8, 8], 4)
    +    batch_spline_spread = torch.ops.nvalchemiops.batch_spline_spread.default(positions_1, charges_1, batch_1, transpose, 5, [8, 8, 8], 4)
    -    zeros = torch.ops.aten.zeros.default([4], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    +    zeros = torch.ops.aten.zeros.default([5], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    ... 4 differing lines in total
```

* nvalchemiops 0.4.1:

```
ewald_reciprocal_space: graphs identical: False
    -    zeros = torch.ops.aten.zeros.default([4], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False)
    +    zeros = torch.ops.aten.zeros.default([5], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False)
    -    zeros_1 = torch.ops.aten.zeros.default([4], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False)
    +    zeros_1 = torch.ops.aten.zeros.default([5], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False)
    -    zeros_2 = torch.ops.aten.zeros.default([4], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    +    zeros_2 = torch.ops.aten.zeros.default([5], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    ... 6 differing lines in total

pme_reciprocal_space: graphs identical: False
    -    batch_spline_spread = torch.ops.nvalchemiops.batch_spline_spread.default(positions_1, charges_1, batch_1, transpose, 4, [8, 8, 8], 4)
    +    batch_spline_spread = torch.ops.nvalchemiops.batch_spline_spread.default(positions_1, charges_1, batch_1, transpose, 5, [8, 8, 8], 4)
    -    zeros = torch.ops.aten.zeros.default([4], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    +    zeros = torch.ops.aten.zeros.default([5], dtype = torch.float64, device = device(type='cuda', index=0), pin_memory = False)
    ... 4 differing lines in total
```

# Environment details

GPU: NVIDIA H100
CUDA version: 12.8
PyTorch version: 2.11.0
NVIDIA Warp version: 1.15.0

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Under `torch.compile`, several batched operations were converting symbolic batch and atom counts into Python integers, while omitted sizing metadata could fall back to `.item()`, causing graph breaks and sometimes synchronizing the device. The patch carries these counts as `torch.SymInt` values through Ewald and PME, including multipole electrostatics, the shared electrostatics/autograd helpers, and the custom spline Warp operator. The patch also emits targeted `FutureWarning`s when DSF, PME mesh setup, reciprocal Ewald, or batched naive, dual-cutoff, and cell-list neighbor construction must infer missing system counts, mesh dimensions, maximum atom counts, or cache and allocation metadata in Python. Callers can supply the named metadata to keep compiled execution graph-safe before the compatibility fallbacks become errors.

Concrete single-system reduction, broadcast, and Ewald atom-range fast paths remain in place, and empty single-system means still return zeros.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #145.

## Changes Made

- Preserve symbolic system and atom counts through reciprocal Ewald, PME, spline, and shared electrostatics helpers.
- Warn during compilation when compatibility metadata is missing and Python inference introduces a graph break; the fallback will become an error in a future release.
- Preserve the concrete single-system reduction, broadcast, and atom-range fast paths, including zero-valued means for empty single-system inputs.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

The warning paths are compatibility fallbacks. Supplying the relevant explicit metadata keeps compiled execution graph-safe.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/interactions/electrostatics/_util.py [hunk 1]
```diff
@@ -352,17 +348,21 @@ def _component_direct_output_deprecation_msg(fn: str, flags: tuple[str, ...]) ->
     )
 
 
+def _is_static_single_system(num_systems: int | torch.SymInt) -> bool:
```
Comment (M-R-Schaefer) [inline nvalchemiops/torch/interactions/electrostatics/_util.py:351]: Several type hints were changed from `int` to `torch.SimInt`. Here the type hint is `int | torch.SymInt`. Should the other type hints also use the union, since they may in principle be called in eager mode as well?

File: nvalchemiops/torch/interactions/electrostatics/pme.py [hunk 2]
```diff
@@ -3151,6 +3144,14 @@ def _select_energy(energy):
...
         if mesh_spacing is None:
             raise ValueError("Either mesh_dimensions or mesh_spacing must be provided")
+        if torch.compiler.is_compiling():
+            warnings.warn(
+                "Missing `mesh_dimensions`; inferring it from `mesh_spacing` and "
+                "`cell` introduces a graph break under torch.compile. This will "
+                "become an error in a future release.",
+                FutureWarning,
+                stacklevel=2,
+            )
```
Comment (M-R-Schaefer) [inline nvalchemiops/torch/interactions/electrostatics/pme.py:3154]: This compile warning block is repeated a few times. Should it be moved to a helper function?

File: nvalchemiops/torch/interactions/electrostatics/pme.py [hunk 3]
```diff
@@ -1680,14 +1676,10 @@ def pme_energy_corrections_with_charge_grad(
...
-            total_charges = charges.to(input_dtype).sum().reshape(1)
-        else:
-            total_charges = torch.zeros(
-                num_systems, dtype=input_dtype, device=raw_energies.device
-            )
-            total_charges.scatter_add_(0, batch_idx, charges.to(input_dtype))
+        total_charges = torch.zeros(
+            num_systems, dtype=input_dtype, device=raw_energies.device
+        )
+        total_charges.scatter_add_(0, batch_idx, charges.to(input_dtype))
```
Comment (M-R-Schaefer) [inline nvalchemiops/torch/interactions/electrostatics/pme.py:1682]: I'm running into a dtype issue in `test_symbolic_make_fx_is_batch_size_independent` as `Tensor.scatter_add_()` requires a `LongTensor` for the index, while `batch_idx` is `int32`. Would
`total_charges.scatter_add_(0, batch_idx.long(), charges.to(input_dtype))`
Be an option?

Comment (dallasfoster) [reply nvalchemiops/torch/interactions/electrostatics/pme.py:1683]: I think this is discovered to be a torch version specific issue. What torch version are you running @M-R-Schaefer?

File: nvalchemiops/torch/interactions/electrostatics/_util.py (hunk 1 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/_util.py:351]: Fixed in 255bc88

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 3 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/pme.py:1683]: torch>=2.8 supports int32 indices.

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 2 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/pme.py]: Fixed in f7760fc.

**Review by dallasfoster — APPROVED** [review]

**Comment by dallasfoster** [comment]
/ok to test [f299ae8](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/150/commits/f299ae8f7188c68b1d7ace6b30251fa14b693f7b)

File: nvalchemiops/torch/interactions/electrostatics/pme.py (hunk 3 above)
Comment (M-R-Schaefer) [reply nvalchemiops/torch/interactions/electrostatics/pme.py:1683]: I see, I was using 2.7.1. Hmm not sure how I ended up with the stale venv. All good then.

**Comment by sklenard** [comment]
Hi @zubatyuk, That’s great, thanks! I’ve tested the fix in my application, and I can confirm it fully resolves #145. 
Woth noting I’ve observed similar issues with other multipole entry points, such as `multipole_electrostatic_energy` and `multipole_particle_mesh_ewald`, which are not addressed by this PR. However, I haven’t investigated these in detail yet, so there might be additional issues to consider.

**Comment by zubatyuk** [comment]
> Hi @zubatyuk, That’s great, thanks! I’ve tested the fix in my application, and I can confirm it fully resolves #145. Woth noting I’ve observed similar issues with other multipole entry points, such as `multipole_electrostatic_energy` and `multipole_particle_mesh_ewald`, which are not addressed by this PR. However, I haven’t investigated these in detail yet, so there might be additional issues to consider.

Thanks for pointing that out. These issues deserve separate PRs.

**Review by dallasfoster — APPROVED** [review]
